ath79: add support for Senao Engenius ECB1200 (DTSI with ECB1750)

mcprat opened the pull request:
FCC ID: A8J-ECB1200

Engenius ECB1200 is an indoor wireless access point with a gigabit ethernet port,
2.4 GHz and 5 GHz wireless, external antennas, and 802.3af PoE.

**Specification:**

  - QCA9557			MIPS SOC, 2.4 GHz (2x2)
  - QCA9882			PCIe on-board, 5 GHz (2x2)
  - AR8035			RGMII, GbE with 802.3af PoE, 25 MHz clock
  - 40 MHz reference clock
  - 16 MB FLASH			25L12845EMI-10G
  - 2x 64 MB RAM		1538ZFZ V59C1512164QEJ25
  - UART at JP1			(unpopulated, RX shorted to ground)
  - 4 external antennas
  - 4 LEDs, 1 button		(power, eth, wifi2g, wifi5g) (reset)

**MAC addresses:**

  MAC Addresses are labeled as ETH and 5GHZ
  U-boot environment has the vendor MAC addresses
  MAC addresses in ART do not match vendor

  eth0	ETH	*:5c	u-boot-env ethaddr
  phy0	5GHZ	*:5d	u-boot-env athaddr
  ----	----	????	art 0x0/0x6

**Installation:**

  Method 1: Firmware upgrade page:

  OEM webpage at 192.168.1.1
  username and password "admin"
  Navigate to "Firmware" page from left pane
  Click Browse and select the factory.bin image
  Upload and verify checksum
  Click Continue to confirm and wait 3 minutes

  Method 2: Serial to load Failsafe webpage:

  After connecting to serial console and rebooting...
  Interrupt uboot with any key pressed rapidly

  run `tftpboot` or `run factory_boot` with initramfs-kernel.bin
  (see TFTP recovery)
  perform a sysupgrade

**Serial Access:**

  the RX line on the board for UART is shorted to ground by resistor R176
  therefore it must be removed to use the console
  but it is not necessary to remove to view boot log
  optionally, R175 can be replaced with a solder bridge short

  the resistors R175 and R176 are next to the UART pinout at JP1

**Return to OEM:**

  If you have a serial cable, see Serial Failsafe instructions

  Unlike most Engenius boards, this does not have a 'failsafe' image
  the only way to return to OEM is TFTP or serial access to u-boot

**TFTP recovery:**

  Unlike most Engenius boards, TFTP is reliable here

  rename initramfs-kernel.bin to 'ap.bin'
  make the file available on a TFTP server at 192.168.1.10
  power board while holding or pressing reset button repeatedly

**Format of OEM firmware image:**

  The OEM software of ECB1200 is a heavily modified version
  of Openwrt Altitude Adjustment 12.09.

  This Engenius board, like ECB1750, uses a proprietary header
  with a unique Product ID. The header for factory.bin is
  generated by the mksenaofw program included in openwrt.

**Note on PLL-data cells:**

  The default PLL register values will not work
  because of the AR8035 switch between
  the SOC and the ethernet port.

  For QCA955x series, the PLL registers for eth0 and eth1
  can be see in the DTSI as 0x28 and 0x48 respectively.
  Therefore the PLL registers can be read from uboot
  for each link speed after attempting tftpboot
  or another network action using that link speed
  with `md 0x18050028 1` and `md 0x18050048 1`.

  However the registers that u-boot sets are not ideal and sometimes wrong...
  the at803x driver supports setting the RGMII clock/data delay on the PHY side.
  This way the pll-data register only needs to handle invert and phase.

  for this board clock invert is needed on the MAC side
  all link speeds functional

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath79: add support for Senao Engenius ECB1200

FCC ID: A8J-ECB1200

Engenius ECB1200 is an indoor wireless access point with a GbE port,
2.4 GHz and 5 GHz wireless, external antennas, and 802.3af PoE.

**Specification:**

  - QCA9557 SOC			MIPS, 2.4 GHz (2x2)
  - QCA9882 WLAN		PCIe card, 5 GHz (2x2)
  - AR8035-A switch		RGMII, GbE with 802.3af PoE, 25 MHz clock
  - 40 MHz reference clock
  - 16 MB FLASH			25L12845EMI-10G
  - 2x 64 MB RAM		1538ZFZ V59C1512164QEJ25
  - UART at JP1			(unpopulated, RX shorted to ground)
  - 4 external antennas
  - 4 LEDs, 1 button		(power, eth, wifi2g, wifi5g) (reset)

**MAC addresses:**

  MAC Addresses are labeled as ETH and 5GHZ
  U-boot environment has the vendor MAC addresses
  MAC addresses in ART do not match vendor

  eth0	ETH	*:5c	u-boot-env ethaddr
  phy0	5GHZ	*:5d	u-boot-env athaddr
  ----	----	????	art 0x0/0x6

**Installation:**

  Method 1: Firmware upgrade page:

  OEM webpage at 192.168.1.1
  username and password "admin"
  Navigate to "Firmware" page from left pane
  Click Browse and select the factory.bin image
  Upload and verify checksum
  Click Continue to confirm and wait 3 minutes

  Method 2: Serial to load Failsafe webpage:

  After connecting to serial console and rebooting...
  Interrupt uboot with any key pressed rapidly

  (see TFTP recovery)
  perform a sysupgrade

**Serial Access:**

  the RX line on the board for UART is shorted to ground by resistor R176
  therefore it must be removed to use the console
  but it is not necessary to remove to view boot log
  optionally, R175 can be replaced with a solder bridge short

  the resistors R175 and R176 are next to the UART pinout at JP1

**Return to OEM:**

  If you have a serial cable, see Serial Failsafe instructions

  Unlike most Engenius boards, this does not have a 'failsafe' image
  the only way to return to OEM is TFTP or serial access to u-boot

**TFTP recovery:**

  Unlike most Engenius boards, TFTP is reliable here

  rename initramfs-kernel.bin to 'ap.bin'
  make the file available on a TFTP server at 192.168.1.10
  power board while holding or pressing reset button repeatedly

  or with serial access:
  run `tftpboot` or `run factory_boot` with initramfs-kernel.bin
  then `bootm` with the load address

**Format of OEM firmware image:**

  The OEM software of ECB1200 is a heavily modified version
  of Openwrt Altitude Adjustment 12.09.

  This Engenius board, like ECB1750, uses a proprietary header
  with a unique Product ID. The header for factory.bin is
  generated by the mksenaofw program included in openwrt.

**Note on PLL-data cells:**

  The default PLL register values will not work
  because of the AR8035 switch between
  the SOC and the ethernet port.

  For QCA955x series, the PLL registers for eth0 and eth1
  can be see in the DTSI as 0x28 and 0x48 respectively.
  Therefore the PLL registers can be read from uboot
  for each link speed after attempting tftpboot
  or another network action using that link speed
  with `md 0x18050028 1` and `md 0x18050048 1`.

  However the registers that u-boot sets are not ideal and sometimes wrong...
  the at803x driver supports setting the RGMII clock/data delay on the PHY side.
  This way the pll-data register only needs to handle invert and phase.

  for this board clock invert is needed on the MAC side
  all link speeds functional

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
boot log
https://gist.github.com/mpratt14/e0762a2a986323c015712a603f584ddc

mpratt14 commented:
ping @svfrd 

This is the sister board of ECB1750 at #1946 

I have a few minor notes on ECB1750 since I think these boards are the same except for RF, and I was wondering if it's possible for you to verify any of this...

that UART is disabled because RX is shorted to ground

I have tested factory.bin, so as long as you have the product ID right, it should be enabled

the led for ethernet is completely controlled by the AR8035 switch, so the lines in the led script should be removed

eth0 node in DTS can be improved, since there have been improvements with the at803x phy driver in linux 5
and understanding the roles of the eth0 PLL register by looking at a datasheet
and also that handling RGMII delay is preferred on the PHY side
you can read about it in my other PR #3505 
the eth0 node in this PR should work perfectly for ECB1750

svfrd commented:
@mpratt14

Thanks for your notes.
 
> that UART is disabled because RX is shorted to ground

On ecb1750 you don't need to remove any resistor.

> I have tested factory.bin, so as long as you have the product ID right, it should be enabled

Nice catch. factory.bin has been defined but never build.
I won't test this one, but it should work.

> the led for ethernet is completely controlled by the AR8035 switch, so the lines in the led script should be removed
> 
> eth0 node in DTS can be improved, since there have been improvements with the at803x phy driver in linux 5
> and understanding the roles of the eth0 PLL register by looking at a datasheet
> and also that handling RGMII delay is preferred on the PHY side
> you can read about it in my other PR #3505
> the eth0 node in this PR should work perfectly for ECB1750 using a build from master

Indeed.. with rgmii-id and the changed pll-data it works.
I have tested it successfully with iperf and the different link speeds.

- 1000Mb Full Duplex

```
root@OpenWrt:~# mii-tool -v eth0
eth0: negotiated 1000baseT-FD flow-control, link ok
  product info: vendor 00:13:74, model 7 rev 2
  basic mode:   autonegotiation enabled
  basic status: autonegotiation complete, link ok
  capabilities: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  link partner: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
root@OpenWrt:~# iperf -s
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size:  128 KByte (default)
------------------------------------------------------------
[  4] local 192.168.1.71 port 5001 connected with 192.168.1.254 port 49158
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.0 sec   540 MBytes   452 Mbits/sec
root@OpenWrt:~#
```

- 100Mb Full Duplex

```
root@OpenWrt:~# mii-tool -v eth0
eth0: 100 Mbit, full duplex, link ok
  product info: vendor 00:13:74, model 7 rev 2
  basic mode:   100 Mbit, full duplex
  basic status: link ok
  capabilities: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
root@OpenWrt:~# iperf -s
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size:  128 KByte (default)
------------------------------------------------------------
[  4] local 192.168.1.71 port 5001 connected with 192.168.1.254 port 49162
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.0 sec   112 MBytes  94.1 Mbits/sec
root@OpenWrt:~# 
```
- 10Mb Full Duplex

```
root@OpenWrt:~# mii-tool -v eth0
eth0: 10 Mbit, full duplex, link ok
  product info: vendor 00:13:74, model 7 rev 2
  basic mode:   10 Mbit, full duplex
  basic status: link ok
  capabilities: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
root@OpenWrt:~# iperf -s
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size:  128 KByte (default)
------------------------------------------------------------
[  4] local 192.168.1.71 port 5001 connected with 192.168.1.254 port 49166
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.3 sec  11.5 MBytes  9.41 Mbits/sec
root@OpenWrt:~#
```

- 10Mb Half Duplex

```
root@OpenWrt:~# mii-tool -v eth0
eth0: 10 Mbit, half duplex, link ok
  product info: vendor 00:13:74, model 7 rev 2
  basic mode:   10 Mbit, half duplex
  basic status: link ok
  capabilities: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
root@OpenWrt:~# iperf -s
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size:  128 KByte (default)
------------------------------------------------------------
[  4] local 192.168.1.71 port 5001 connected with 192.168.1.254 port 49166
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.3 sec  11.5 MBytes  9.40 Mbits/sec
root@OpenWrt:~#
```

> also for ECB1200 the first MAC on label is for ethernet, it would be the same for ECB1750 maybe?

You are right. I'll fix and test the label tomorrow.


A common dtsi would be correct for both routers.
You can cherry pick my last commits 
https://github.com/svfrd/openwrt/commits/ecb1750.master

boot log: https://gist.github.com/svfrd/54d30fd21236a0570560932905a1f266

mpratt14 commented:
@svfrd Nice, I'll start the dtsi commit

another small thing, for some reason on both boards Openwrt says the SOC is QCA9558 but for my ECB1200 it is actually QCA9557 if you look at the chip, for now I assume this is the same for ECB1750

mcprat committed:
ath79: create common DTSI for ECB1200 and ECB1750

These boards are sister boards

exactly the same hardware except that ECB1200 has:
 - QCA9557
 - 2 RF circuits/antennas per band instead of 3
 - a resistor blocking UART RX line

Tested-by: sven friedmann <sf.openwrt@okay.ms>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath79: enable factory.bin and adjust profile of ECB1750

factory.bin was not tested for ECB1750...
but it was tested on it's sister board ECB1200

The product ID for the header can be verified by inspecting
the header of OEM images, or in the u-boot environment.

Also:

  - the LAN LED is controlled directly by the AR8035 switch
  - the labelled (first increment) MAC for both is ethaddr (eth0)
  - list packages in alphabetical order
  - use default sysupgrade.bin recipe

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@svfrd OK it should be good now, thanks for the information

adschm commented on target/linux/ath79/dts/qca9557_engenius_ecb1200.dts:
```diff
@@ -0,0 +1,143 @@
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
```
num-cs can probably be dropped ... (same for your other ath79 devices)

mpratt14 commented:
@adschm on all my open PRs I have:

dropped num-cs
updated commit message
rebased

mpratt14 commented:
@svfrd label MAC was messed up by both of us, fixed now

svfrd reviewed (commented):
you change unintentionally the ecb1750 profile (factory.bin) in this commt [1].. pls fix.

[1] https://github.com/openwrt/openwrt/pull/3630/commits/ec10d78e3ebfd20b58b54d32efd90624469e6425

svfrd reviewed (approved):
tested and works on ecb1750!

mpratt14 commented:
@svfrd thanks for testing

the changes to ecb1750 profile are in the last commit only though, not sure what you're seeing...

mpratt14 commented:
rebased and commit message edit

mpratt14 commented:
ready for final review whenever you have time

mpratt14 commented:
commit message edit again

mpratt14 commented:
@adschm rebased due to your recent commit, ready for comments 👍🏽

svfrd reviewed (approved):
successfully tested on ecb1750.

bootlog: https://gist.github.com/svfrd/15e0a2ba4bed68c820a442809e086870

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
